lib/compiler.js

```diff
@@ -538,3 +531,43 @@ function resolveRelativePaths(relativePaths, appRootDir) {
     }
   });
 }
+
+function isPreferredExtension (filename) {
+  var includeExtensions = require.extensions;
+  var excludeExtensions = {};
+  excludeExtensions[FILE_EXTENSION_JSON] = FILE_EXTENSION_JSON;
+
+  var ext = path.extname(filename);
+  return (ext in includeExtensions) && !(ext in excludeExtensions);
+}
+
+function fixFileExtension (filepath) {
+  var results = [];
+  var otherFile;
+
+  /* Prefer coffee scripts over json */
+  if (isPreferredExtension(filepath)) return filepath;
+
+  var basename = path.basename(filepath, FILE_EXTENSION_JSON);
+  var sourceDir = path.dirname(filepath);
+
+  var files = tryReadDir(sourceDir);
+  files.forEach(function(f) {
+    otherFile = path.resolve(sourceDir, f);
+
+    var stats = fs.statSync(otherFile);
+    if (stats.isFile()) {
+      var otherFileExtension = path.extname(f);
+      if (otherFileExtension != FILE_EXTENSION_JSON &&
+        path.basename(f, otherFileExtension) == basename) {
+        results.push(otherFile);
+      }
+    }
+  });
+  return (results.length > 0 ? results[0] : filepath);
+}
```

Comment:
This code is very similar to what we have in [loadModuleDefinition()](https://github.com/PradnyaBaviskar/loopback-boot/blob/8e9c43c16b029ee534123bf2e0b37145c335431e/lib/compiler.js#L358-L385). Can you please refactor these two methods to share as much code as possible?

I also think the approach in `loadModuleDefinition` is better because it calls `tryReadDir` only once, while `fixFileExtension` is calling it for every file being resolved.